ethereum/core/src/main/java/tech/pegasys/teku/core/DepositMerkleTree.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.core;
+
+import static java.lang.StrictMath.toIntExact;
+import static tech.pegasys.teku.util.config.Constants.DEPOSIT_CONTRACT_TREE_DEPTH;
+import static tech.pegasys.teku.util.config.Constants.MAX_DEPOSITS;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.List;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.datastructures.operations.Deposit;
+import tech.pegasys.teku.datastructures.operations.DepositWithIndex;
+import tech.pegasys.teku.datastructures.state.BeaconState;
+import tech.pegasys.teku.datastructures.util.MerkleTree;
+import tech.pegasys.teku.datastructures.util.OptimizedMerkleTree;
+import tech.pegasys.teku.ssz.SSZTypes.SSZList;
+
+public class DepositMerkleTree {
+  private final MerkleTree depositMerkleTree = new OptimizedMerkleTree(DEPOSIT_CONTRACT_TREE_DEPTH);
+  private final NavigableMap<UnsignedLong, DepositWithIndex> depositNavigableMap = new TreeMap<>();
+
+  public void addDeposit(DepositWithIndex deposit) {
+    depositNavigableMap.put(deposit.getIndex(), deposit);
+    depositMerkleTree.add(deposit.getData().hash_tree_root());
+  }
+
+  public SSZList<Deposit> getDeposits(BeaconState state) {
```

Comment:
I agree, the only reason I did that was because the `DepositProvider` has a reader specific API. It only accepts through `DepositsFromBlockEvents`. For testing, it would be better to have a class that accepts through only deposits as well.